website/static/worker.js

```diff
@@ -30,6 +30,25 @@ for (const file in parsersLocation) {
   }
 }
 
+const docExplorerPlugin = {
+  parsers: {
+    "doc-explorer": {
+      parse: (text) =>
+        new Function(
```

Comment:
How about use a js parser to parse input , and check the AST first, just to be safe?

[1](https://deploy-preview-10183--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBuAOlTBzAThAVwAcAKAbUwAJL0QSAqWgGisoEsoATBGc16gDYc4LKNWq0ACrjgwYbOLiTN+lIVBGraAeSIc20AIbxOLEKPFrhF8bQDCEbmZuDrWkADEIuALbH4uCpilAC6AJQ26prBtPRhQeHmIBBE8tAAzsighrj4AO6SOQiZKIYCeYYAnplMIABGuIZgANayAMpETRzYyDC4BCIgcD51cJzcnAAyhlDYBIbYcF6+-t3IIIYEMBBJABYwPgIA6rts8OmdYHBtxWdsAG5nletg6TUgHOmKMNILfsgAMzKX1qACt0gAPABCjRa7UMPjgk2EgOBg3BELa3QEcAAigQIPBUQIQSBOrgvrh1pwIGAALRwCFEATeRRJIi4DgwI5sTgwXbIADMAAZahyIF8jo0iOsOXBKfc4EkAI4E+DSFIlDbpOkaMZjJIyVVsGS-bD-JBAkmDL4+Ni9fo27F4tVKy1o2owQx1Hl8gVIABMnsabHU2AcPgtQ3SAFYkgQvgAVb0lK2k+4DACSXB4bTAnNSAEEuG0YJUccSvgBfKtAA)

[2](https://deploy-preview-10183--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBnOAbAZgOjBidACgEoQAaECABxgEtpVlQBDAJzYgHcAFdhJihYYuLAJ5NKAIzYswAazgwAytTl0oAc2Qw2AVziU4AWylwAJuYsAZFlr0tNcAGIQ2xljHpbkIFnpgIChAACxhjDAB1ELp4VDUwOGUBWLoAN1ixXzBUSRANdDYYHllND2QsYXRKACtUAA8AIVkFJWUWYzhrDTgKqsMQOvrlDU0MOABFPQh4PoxqkDU2Qt9zCDAAWjh66gI2ODZg6jYNGEi6cxgQ5ABmAAZKY8I4SNlqX2O4QrTeygBHabwEo0QR+VAbKBwCwWYL7AF0fYlRzlJCVeYDdDGOg6fQY0bjKYzXqo-qUGAsKTnS7XJAAJjJsjoGFGAGEIMYUSAvgBWYJ6dAAFQpgjRCzSBgAklArLBlGATrQAILS5QwMTjOboAC+WqAA)

[3](https://deploy-preview-10183--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBtAOlABFgyjAJwEsoBzACkxywDM4YwALcgckZhgAcBnJAej6kIEUgBs4AOkgBbFgEoqWBVAC6IADQgInGEWjdkoAIYECEAO4AFEwgMojo80YCeBzQCMCRsAGt6uTm8SUmRCAFc4TThpdzgAEzj4gBkjMjCjUjgAMQgCaSMOYOQQIzCYCA0QdmlRAHVGInhuQLA4XFtGogA3RudisG43EBJuOAIYSy9SfOQaB1HNACtuAA8AIS9ffyNpOCSSOFn5yJBlldxg8QBFMIh4I9EFkECCUYJiuIgwAFo4Fc5RLkxpVOMRYLUiHEYIxkABmAAMmlBEFGtS8nGKoLgby6h00AEdbvBJto7CVuN8oHB4vFKgQ4ISiPTJhkZkg5o8TqNpERQgQIppuJc4Dc7od2cdNDAjO4IVCYUgAExSrxEUTBADCEGkbJA2IArJUwqMACoyuwcp5dCIASSgiVguDAxB0AEF7fhnOIHqMAL6+o)

Reply:
Building a sandbox (a good one) is a tricky problem, but do we really need to worry about that in this case? Your examples are harmless. `fetch` can do only same-domain requests. As long as the site is static, it's not a threat. Syntax errors work as expected. `self.close()` doesn't look like a big deal to me. There are also infinite loops... We can use `setTimeout` and terminate the worker if it doesn't respond for too long, but is it really needed? Do you see a way for malicious users to abuse this somehow?

Reply:
> Do you see a way for malicious users to abuse this somehow?

I didn't try much, but I think it's not hard to verify input, simple traverse only allow those builders to call, and string, array, object, literals and `Symbol`, `Symbol.for` , enough?

Reply:
I'd rather have the full power of JS there. It's a development tool, and development needs flexibility. E.g., to test issues with reference equality of different subbranches of the doc, variables are needed (with everything wrapped in an IIFE).

Reply:
For now I'm going to add two things:
* CSP headers via netlify.toml
* Protection against infinite loops
Both seem to be generally useful, not only for doc explorer.

Reply:
Still think it's dangerous, [URL.createObjectURL](https://deploy-preview-10183--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAKVBKABAXgHxbAA6UWWAlrlgEwlkDuAFuQDZxarkC0XGdWkKAGcIbAHQsIAc1T8yAVQBKAGTFgATnACG8APIAjAFZwwMJctRQ49LACFJ+1AG05ZLERAAHTTBjk46u4gYpqe2jCoAIwADHykbgC6ADSEMACeYUhYAOTwAB4wAPSeLFrkUNkAvhhxZLVYPgCu6qTAlSTVmCBJIBCeftBCyKBa6uoQ9AAKowhDKFos9FppQz366lpgANZwMADKnpvlUsgw6o1wPXAAtvpwACb3D8paUFKNWlJwAGIQ6tc6PxvZAgLSNGAQbogRgwa4sADqzHgQkOYDge1m5D8ADcsWkQWAhKsQOUhAEYJMNlIAcgAGYLMk9QxCPK2DbbXZ7LTXODKcpwOkMy4gZl5PbHNgARUaEHggpYjK8ozJ6hB9wgYC4cDyJT+ASh3nKMHh5HuMEYyAAzNEet4IGT4RtPCDvHAVdiBT0AI4y+CUvpzUFCLhWB4PKGaH3kTSUz40pD0hXCsnXcinc7JiVwaWygUJoU9GBafQms0WpDUQsbVjHADCEGu8ZAboArFDGmSACrFuaJxXYi4ASSgT1gew05H6AEER3t0mx5WTKpUgA)

Reply:
I'm going to add a "click to run" button when the playground is opened with `--parser doc-explorer` and a non-default input. It can't be considered dangerous if it's not executed automatically.

I didn't understand the danger of `createObjectURL` though. What's the problem with it?

Reply:
Crash browse, slow down system, use disk(maybe browser has limitation).

Reply:
Anyway, if it's the user who decides to run the code, all of these aren't a problem. Only automatic execution can be a problem.

Reply:
I guess that's better.